toolkit/offchain/src/register.rs

```diff
@@ -0,0 +1,293 @@
+#![allow(dead_code)]
+
+use crate::csl::{InputsBuilderExt, TransactionBuilderExt, TransactionContext};
+use crate::plutus_script::PlutusScript;
+use cardano_serialization_lib::{
+	BigNum, DataCost, ExUnits, JsError, MinOutputAdaCalculator, Transaction, TransactionBuilder,
+	TransactionOutputBuilder, TxInputsBuilder,
+};
+use ogmios_client::types::OgmiosUtxo;
+use partner_chains_plutus_data::registered_candidates::block_producer_registration_to_plutus_data;
+use sidechain_domain::BlockProducerRegistration;
+
+pub fn register_tx(
+	validator: &PlutusScript,
+	block_producer_registration: &BlockProducerRegistration,
+	input_utxo: &OgmiosUtxo,
+	own_registration_utxos: &[OgmiosUtxo],
+	ctx: &TransactionContext,
+	validator_redeemer_ex_units: ExUnits,
+) -> Result<Transaction, JsError> {
+	let config = crate::csl::get_builder_config(ctx)?;
+	let mut tx_builder = TransactionBuilder::new(&config);
+
+	{
+		let mut inputs = TxInputsBuilder::new();
+		for own_registration_utxo in own_registration_utxos {
+			inputs.add_script_utxo_input(
+				own_registration_utxo,
+				validator,
+				validator_redeemer_ex_units.clone(),
+			)?;
+		}
+		inputs.add_key_inputs(&[input_utxo.clone()], &ctx.payment_key_hash())?;
+		tx_builder.set_inputs(&inputs);
+	}
+
+	{
+		let amount_builder =
+			TransactionOutputBuilder::new().with_address(&ctx.payment_address()).next()?;
+		let output = amount_builder.with_coin(&BigNum::zero()).build()?;
+		let min_ada = MinOutputAdaCalculator::new(
+			&output,
+			&DataCost::new_coins_per_byte(
+				&ctx.protocol_parameters.min_utxo_deposit_coefficient.into(),
+			),
+		)
+		.calculate_ada()?;
+		let output = amount_builder.with_coin(&min_ada).build()?;
+		tx_builder.add_output(&output)?;
+	}
```

Comment:
I'm not saying it is an error here, but why do we need to have payment_address output? Doesn't balancing add change if required?
I'm wondering if I should have added it to d-param and permissioned-candidates.

I think that with this output, tx requires over `2 x min_ada + C`, where `C` is the cost of datum at validator address.
OTOH without this address and wallet that has less than that, say `min_ada + C + D` will lose `D`, because it is not enough for keeping change UTXO.